src/transform/src/fusion/join.rs

```diff
@@ -50,85 +50,152 @@ impl Join {
-                    // Add all of the inputs.
-                    for input in inputs.drain(..) {
-                        new_columns += input.arity();
-                        new_inputs.push(input);
+                    _ => {
+                        // Retain the input.
+                        join_builder.add_input(input);
                     }
-                } else {
-                    // Retain the input.
-                    new_columns += input.arity();
-                    new_inputs.push(input);
                 }
             }
 
-            // Filter join identities out of the inputs.
-            // The join identity is a single 0-ary row constant expression.
-            new_inputs.retain(|input| {
-                if let MirRelationExpr::Constant {
-                    rows: Ok(rows),
-                    typ,
-                } = &input
-                {
-                    !(rows.len() == 1 && typ.column_types.len() == 0 && rows[0].1 == 1)
-                } else {
-                    true
-                }
-            });
-
-            new_equivalences.extend(equivalences.drain(..));
-            expr::canonicalize::canonicalize_equivalences(&mut new_equivalences);
-
-            *inputs = new_inputs;
-            *equivalences = new_equivalences;
-            *demand = None;
-            *implementation = expr::JoinImplementation::Unimplemented;
-
-            // If `inputs` is now empty or a singleton (without constraints),
-            // we can remove the join.
-            match inputs.len() {
-                0 => {
-                    // The identity for join is the collection containing a single 0-ary row.
-                    *relation = MirRelationExpr::constant(vec![vec![]], RelationType::empty());
-                }
-                1 => {
-                    // if there are constraints, they probably should have
-                    // been pushed down by predicate pushdown, but .. let's
-                    // not re-write that code here.
-                    if equivalences.is_empty() {
-                        *relation = inputs.pop().unwrap();
+            *relation = join_builder.build();
+        }
+    }
+}
+
+/// Helper builder for fusing the inputs of nested joins into a single Join expression.
+struct JoinBuilder {
+    inputs: Vec<MirRelationExpr>,
+    equivalences: Vec<Vec<MirScalarExpr>>,
+    num_columns: usize,
+    /// Predicates that will be evaluated on top of the join, if any.
+    predicates: Vec<MirScalarExpr>,
+}
+
+impl JoinBuilder {
+    fn new(equivalences: &mut Vec<Vec<MirScalarExpr>>) -> Self {
+        Self {
+            inputs: Vec::new(),
+            equivalences: equivalences.drain(..).collect(),
+            num_columns: 0,
+            predicates: Vec::new(),
+        }
+    }
+
+    fn add_input(&mut self, input: MirRelationExpr) {
+        self.num_columns += input.arity();
+        self.inputs.push(input);
+    }
+
+    fn add_inputs<I>(
```

Comment:
Nit: I'm in favor of `add_subjoin` instead of `add_inputs` because the name `add_inputs` makes me think it is a convenience method for
```
        for input in inputs {
            self.add_input(input);
        }
```

Reply:
good point